torch/csrc/autograd/python_variable.cpp

```diff
@@ -525,16 +525,83 @@ static PyObject* THPVariable_fix_weakref(PyObject* self, PyObject* noargs) {
   Py_RETURN_NONE;
 }
 
+// Maps the given PyObject callable over a vector of items. The wrapper /
+// unwrapper args define how to convert between PyObject and item type.
+template <typename T, typename Wrapper, typename Unwrapper>
+static std::vector<T> map_py_func(
+    PyObject* func,
+    const std::vector<T>& items,
+    Wrapper wrapper,
+    Unwrapper unwrapper) {
+  std::vector<T> new_items;
+  new_items.resize(items.size());
+  std::transform(
+      items.begin(), items.end(), new_items.begin(), [&](const T& item) {
+        PyObject* res = PyObject_CallOneArg(func, wrapper(item));
```

Comment:
You are definitely leaking this res.
I think you want to either
- Move all this code to use pybind constructs
- If not, use THPObjectPtr to manage lifetime
- Check the doc for every CPython API  you use to know if it returns borrowed or new reference and if it steals reference to the args.

Reply:
thanks good catch! I'm definitely not too familiar with these APIs. going to dig into them and clean this up

Reply:
Fixed this, much nicer now